libload.asm

```diff
@@ -51,25 +49,39 @@ libnameptr                 = pixelShadow+42 ; pointer to name of library to extr
 ; macro definitions
 define lib_byte            $C0		; library signifier byte
 define jp_byte             $C3		; byte for 'jp' opcode
+define libmagic1           $C0		; library magic byte 1
+define libmagic2           $C1		; library magic byte 2
+define libmagic1alt        $BF		; alternate library magic byte 1
+define libmagic2alt        $FE		; alternate library magic byte 2
 define asmflag             $22		; flag storage
 
 define prevextracted       0
 define foundprgmstart      1
 define keeplibinarc        2
 
-macro relocate location
-	olocation = $
-	org location
-	rlocation = location
+macro relocate? name, address*
+	name.source? := $
+	name.destination? := address
+	org name.destination?
+	macro end?.relocate?
+		name.length? := $ - name.destination?
+		org name.source? + name.length?
+		purge end?.relocate?
+	end macro
 end macro
 
-macro endrelocate
-	org $-rlocation+olocation
-end macro
-
-org 0					; base location
-
-_libload:				; this code executes in the archive (entered with jp (hl)
+; The alternate magic bytes were added to allow LibLoad to present itself as a
+; library (initially just for the version check, but LibLoad could expose
+; functions in the future as well). For compatibility, the LibLoad invocation
+; process starts execution from the start of this header, so it needs to be
+; harmless when executed. With these magic bytes, the header is decoded as:
+;	cp	a,a
+;	cp	a,VERSION_MAJOR*10+VERSION_MINOR
+library 'LibLoad', VERSION_MAJOR*10+VERSION_MINOR, libmagic1alt, libmagic2alt
+
+; We *are* the relocator, so we can't use relocations here. Set origin to 0
+; (shifted by library stuff before this) and perform any relocations manually.
+relocate _libload, 3			; FIXME magic origin
```

Comment:
Any way to derive this origin value properly? It should be equal to the size of all the library stuff that precedes it.

Reply:
That would work for now, as LibLoad exports no symbols and requires no relocations. But if exported symbols are added in the future, relocations will be needed, so entirely disabling relocations seems like a suboptimal idea. I'd much rather just have a way of referencing/deriving the header size and using the current structure.

Reply:
Fixed by exporting related macros.